src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs

```diff
@@ -373,73 +401,76 @@ public async Task StopAsync(CancellationToken cancellationToken)
 
         public async Task RestartHostAsync(CancellationToken cancellationToken)
         {
-            using (_metricsLogger.LatencyEvent(MetricEventNames.ScriptHostManagerRestartService))
+            if (!ShutdownRequested)
             {
-                // Do not invoke a restart if the host has not yet been started. This can lead
-                // to invalid state.
-                if (!_hostStarted.IsCompleted)
+                using (_metricsLogger.LatencyEvent(MetricEventNames.ScriptHostManagerRestartService))
                 {
-                    _logger.RestartBeforeStart();
-                    await _hostStarted;
-                }
+                    // Do not invoke a restart if the host has not yet been started. This can lead
+                    // to invalid state.
+                    if (!_hostStarted.IsCompleted)
+                    {
+                        _logger.RestartBeforeStart();
+                        await _hostStarted;
+                    }
 
-                _logger.EnteringRestart();
+                    _logger.EnteringRestart();
 
-                // If anything is mid-startup, cancel it.
-                _startupLoopTokenSource?.Cancel();
-                foreach (var startupOperation in ScriptHostStartupOperation.ActiveOperations)
-                {
-                    _logger.CancelingStartupOperationForRestart(startupOperation.Id);
-                    try
+                    // If anything is mid-startup, cancel it.
+                    _startupLoopTokenSource?.Cancel();
+                    foreach (var startupOperation in ScriptHostStartupOperation.ActiveOperations)
                     {
-                        startupOperation.CancellationTokenSource.Cancel();
+                        _logger.CancelingStartupOperationForRestart(startupOperation.Id);
+                        try
+                        {
+                            startupOperation.CancellationTokenSource.Cancel();
+                        }
+                        catch (ObjectDisposedException)
+                        {
+                            // This can be disposed at any time.
+                        }
                     }
-                    catch (ObjectDisposedException)
+
+                    try
                     {
-                        // This can be disposed at any time.
-                    }
-                }
+                        await _hostStartSemaphore.WaitAsync();
```

Comment:
>await _hostStartSemaphore.WaitAsync(); [](start = 24, length = 38)

Realized _hostStartSemaphore already seems to be doing the job of allowing one restart to happen at a time.